core/src/main/java/com/sequenceiq/cloudbreak/service/rdsconfig/AbstractRdsConfigProvider.java

```diff
@@ -42,66 +41,97 @@
+     *
+     * @param  stack   stack for cluster
+     * @param  cluster cluster to associate database with
+     * @return         all RDSConfig objects for the cluster, potentially including a new one for
+     *                 this provider's database
+     */
     public Set<RDSConfig> createPostgresRdsConfigIfNeeded(Stack stack, Cluster cluster) {
         Set<RDSConfig> rdsConfigs = rdsConfigService.findByClusterId(cluster.getId());
         rdsConfigs = rdsConfigs.stream().map(c -> rdsConfigService.resolveVaultValues(c)).collect(Collectors.toSet());
         if (isRdsConfigNeeded(cluster.getBlueprint())
                 && rdsConfigs.stream().noneMatch(rdsConfig -> rdsConfig.getType().equalsIgnoreCase(getRdsType().name()))) {
+            RDSConfig newRdsConfig;
             if (dbServerConfigurer.isRemoteDatabaseNeeded(cluster)) {
-                RDSConfig rbRdsConfig = dbServerConfigurer.getRdsConfig(stack, cluster, getDb(), getRdsType());
-                rdsConfigs = populateRdsConfig(rdsConfigs, stack, cluster, rbRdsConfig);
+                newRdsConfig = dbServerConfigurer.createNewRdsConfig(stack, cluster, getDb(), getDbUser(), getRdsType());
             } else {
                 LOGGER.debug("Creating postgres Database for {}", getRdsType().name());
-                rdsConfigs = createPostgresRdsConf(stack, cluster, rdsConfigs, getDbUser(), getDbPort(), getDb());
+                newRdsConfig = createNewRdsConfig(stack, cluster, getDb(), getDbUser(), getDbPort());
             }
+            rdsConfigs = populateNewRdsConfig(rdsConfigs, stack, cluster, newRdsConfig);
         }
         return rdsConfigs;
     }
 
-    private Set<RDSConfig> populateRdsConfig(Set<RDSConfig> rdsConfigs, Stack stack, Cluster cluster, RDSConfig rdsConfig) {
+    private Set<RDSConfig> populateNewRdsConfig(Set<RDSConfig> rdsConfigs, Stack stack, Cluster cluster, RDSConfig rdsConfig) {
         rdsConfig = rdsConfigService.createIfNotExists(stack.getCreator(), rdsConfig, stack.getWorkspace().getId());
-        if (rdsConfigs == null) {
-            rdsConfigs = new HashSet<>();
-        }
         rdsConfigs.add(rdsConfig);
         cluster.setRdsConfigs(rdsConfigs);
         clusterService.save(cluster);
 
         return rdsConfigs;
     }
 
-    private Set<RDSConfig> createPostgresRdsConf(Stack stack, Cluster cluster,
-            Set<RDSConfig> rdsConfigs, String dbUserName, String dbPort, String dbName) {
+    /**
+     * Creates an RDSConfig object for a specific database.
+     *
+     * @param  stack   stack for naming purposes
+     * @param  cluster cluster to associate database with
+     * @param  dbName  database name
+     * @param  dbUser  database user
+     * @param  dbPort  port for database connections (through gateway)
+     * @return         RDSConfig object for database
+     */
+    private RDSConfig createNewRdsConfig(Stack stack, Cluster cluster, String dbName, String dbUserName, String dbPort) {
```

Comment:
NIT: createNew... seems redundant.

Reply:
That was on purpose, remembering how confused I was whether an RDSConfig was being reused or not. :D